src/rgw/rgw_sync_module_pubsub_rest.cc

```diff
@@ -630,97 +661,281 @@ static int notif_bucket_path(const string& path, string *bucket_name)
+    if (bucket_info.owner != id) {
+      ldout(s->cct, 1) << "user doesn't own bucket, cannot create notification" << dendl;
       return -EPERM;
     }
     return 0;
   }
+
   void pre_exec() override {
     rgw_bucket_object_pre_exec(s);
   }
-  void execute() override;
 
+  RGWOpType get_type() override { return RGW_OP_PUBSUB_NOTIF_CREATE; }
+  uint32_t op_mask() override { return RGW_OP_TYPE_WRITE; }
+};
+
+
+// command (ceph specific): PUT /notification/bucket/<bucket name>?topic=<topic name>
+class RGWPSCreateNotif_ObjStore_Ceph : public RGWPSCreateNotifOp {
+private:
+  std::string topic_name;
+  std::set<std::string, ltstr_nocase> events;
+
+  int get_params() override {
+    bool exists;
+    topic_name = s->info.args.get("topic", &exists);
+    if (!exists) {
+      ldout(s->cct, 1) << "missing required param 'topic'" << dendl;
+      return -EINVAL;
+    }
+
+    string events_str = s->info.args.get("events", &exists);
+    if (exists) {
+      get_str_set(events_str, ",", events);
+    }
+    return notif_bucket_path(s->object.name, bucket_name);
+  }
+
+public:
   const char* name() const override { return "pubsub_notification_create"; }
-  virtual RGWOpType get_type() override { return RGW_OP_PUBSUB_NOTIF_CREATE; }
-  virtual uint32_t op_mask() override { return RGW_OP_TYPE_WRITE; }
-  virtual int get_params() = 0;
+  void execute() override;
 };
 
-void RGWPSCreateNotifOp::execute()
+void RGWPSCreateNotif_ObjStore_Ceph::execute()
 {
-  op_ret = get_params();
-  if (op_ret < 0) {
-    return;
-  }
+  ups = std::make_unique<RGWUserPubSub>(store, s->owner.get_id());
 
-  ups = make_unique<RGWUserPubSub>(store, s->owner.get_id());
   auto b = ups->get_bucket(bucket_info.bucket);
   op_ret = b->create_notification(topic_name, events);
   if (op_ret < 0) {
-    ldout(s->cct, 20) << "failed to create notification, ret=" << op_ret << dendl;
+    ldout(s->cct, 1) << "failed to create notification, ret=" << op_ret << dendl;
     return;
   }
 }
 
+namespace {
+// conversion functions between S3 and GCP style event names
+std::string s3_to_gcp_event(const std::string& event) {
+  if (event == "s3:ObjectCreated:*") {
+    return "OBJECT_CREATE";
+  }
+  if (event == "s3:ObjectRemoved:*") {
+    return "OBJECT_DELETE";
+  }
+  return "UNKNOWN_EVENT";
+}
+std::string gcp_to_s3_event(const std::string& event) {
+  if (event == "OBJECT_CREATE") {
+    return "s3:ObjectCreated:";
+  }
+  if (event == "OBJECT_DELETE") {
+    return "s3:ObjectRemoved:";
+  }
+  return "UNKNOWN_EVENT";
+}
+
+// return a unique topic by prefexing with the notification name: <notification>_<topic>
+std::string topic_to_unique(const std::string& topic, const std::string& notification) {
+  return notification + "_" + topic;
+}
+
+// extract the topic from a unique topic of the form: <notification>_<topic>
+std::string unique_to_topic(const std::string& unique_topic, const std::string& notification) {
```

Comment:
@cbodley i know. i left it there for "completeness" - since I already have the inverse function (which is used). will add #pragma to avoid the warning